woocommerce/woocommerce-admin
# Reports sync refactor

Fixes #2765

Refactors the action scheduling logic for orders and customers to make adding new data easier and more maintainable.  It does this by separating shared functionality for schedulers into `SchedulerTraits` and import functionality into `ImportScheduler`.

**SchedulerTraits**

This trait facilitates adding batch scheduling, action clearing, and action dependencies to any class.  This allows each scheduler to be declarative about its actions and dependencies so that other controllers don't need to be knowledgeable about all of its actions.  Controllers importing this trait should describe 3 items:

* `$name` - Unique identifying name for the scheduler.
* `get_actions()` - Returns an array of key => value pairs for the action name and the hook for that action respectively.
* `get_dependencies()` - Returns an array of key => values pairs for action names and a dependency hook to look for.

It also handles a bunch of trivial functions, such as suffixing actions with the scheduler name to help prevent collision between actions, rescheduling blocked/dependent actions, and flattening arguments sent to the scheduler.

**ImportScheduler**

This abstract class houses functionality that handles batch importing and deletion specific to importer schedulers.

In addition to basic scheduling traits, classes that inherit functions from `ImportScheduler` just need to describe a few methods to work:
* `get_items()` - This retrieves the original item IDs and total count.
* `get_total_imported()` - Used to determine how many items need to be deleted.
* `import()` - The function that action scheduler will run when a single import is being processed, passed the ID from `get_items()`.
* `delete()` - The function that action scheduler will run when deleting a batch of items, passed the number of items to delete.

### Detailed test instructions:

1. Go to Analytics->Settings.
2. Test importing and deleting data.
3. Make sure data is imported correctly.
4. Check to make sure no oddities (unnecessary additional scheduled actions) occur with scheduled actions in `wp_posts`.
